### PR TITLE
Read creds from env if secretFromSystem is true

### PR DESCRIPTION
**Pull request by sanyu**
This PR https://github.com/virtuozzo/external-storage/pull/35 adds a new flag to allow the user to store the Secret object in the kube-system namespace. 
If this flag is enabled in the provisioner, _ploop-flexvol_ will not have the credentials added to _kubernetes.io/secret/clusterName_ / _kubernetes.io/secret/clusterPassword_, so in this case, the credentials are read from the environment.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -309,9 +323,27 @@ func (p Ploop) Mount(target string, options map[string]string) (*flexvolume.Resp
...
+	if err != nil {
+		glog.Fatalf("Could not stat parent dir of %s",mount)
 		return nil, err
 	}
+	if stat.Sys().(*syscall.Stat_t).Dev != rootStat.Sys().(*syscall.Stat_t).Dev {
+		if err := syscall.Unmount(mount, 0); err != nil {
+			glog.Infof("umount failed for %s",mount)
+			return nil, err
+		}
+	} else {
```
Comment (avagin) [inline main.go]: Why can't we check a error from umount()?
```
err := syscall.Unmount(mount, 0)
if err == syscall.EINVAL {
    glog.Infof("%s isn't a mount point",mount)
    return nil, err
} elif err != nil {
    return nil, err
}
```

File: main.go [hunk 2]
```diff
@@ -229,6 +239,10 @@ func (p Ploop) Mount(target string, options map[string]string) (*flexvolume.Resp
 	if options["kubernetes.io/readwrite"] == "ro" {
 		readonly = true
 	}
+	if options["secretFromSystem"] == "true" {
+		options["kubernetes.io/secret/clusterName"] = os.Getenv("clusterName")
+		options["kubernetes.io/secret/clusterPassword"] = os.Getenv("clusterPassword")
```
Comment (avagin) [inline main.go]: I think we should use more unique names for these variables to be sure that we don't intersect with someone else.
For example, they can be kubeVstorageClusterName and kubeVstorageClusterPassword.

BTW: how are you going to set these variables?

File: main.go [hunk 3]
```diff
@@ -10,6 +10,8 @@ import (
 	"os/exec"
 	"path/filepath"
 	"syscall"
+	spath "path"
```
Comment (avagin) [inline main.go]: it is better to use filepath instead of path, and it is already imported

File: main.go (hunk 2 above)
Comment (sanyu) [reply main.go]: No problem...I'll rename those variables.
Currently we are setting these variables through the kubelet's environment using systemd
```
# cat /etc/systemd/system/kubelet.service.d/15-ploop.conf
[Service]
EnvironmentFile=/etc/sysconfig/ploop-flexvol
```
```
# cat /etc/sysconfig/ploop-flexvol
clusterName="base64encodedClusterName"
clusterPassword="base64encodedPassword"
workingDir="/vstorage"
```

**Comment by avagin** [comment]
LGTM, thanks.

@dmission Could you take a look at this pr?

File: main.go [hunk 4]
```diff
@@ -136,6 +136,14 @@ func (p Ploop) Init() (*flexvolume.Response, error) {
 	}, nil
 }
 
+func (p Ploop) realWorkingDir(options map[string]string) string {
+	realDir := workingDir
+	if options["secretFromSystem"] == "true" {
```
Comment (dmission) [inline main.go]: workingDir is not a secret, thus it is not obvious that it is managed by 'secretFromSystem'. Let's name this option something like 'optionsFromSystem'.

File: main.go [hunk 5]
```diff
@@ -229,6 +237,10 @@ func (p Ploop) Mount(target string, options map[string]string) (*flexvolume.Resp
 	if options["kubernetes.io/readwrite"] == "ro" {
 		readonly = true
 	}
+	if options["secretFromSystem"] == "true" {
+		options["kubernetes.io/secret/clusterName"] = os.Getenv("kubeVstorageClusterName")
+		options["kubernetes.io/secret/clusterPassword"] = os.Getenv("kubeVstorageClusterPassword")
+	}
```
Comment (dmission) [inline main.go]: Is there any need to modify a function's argument ("options")? I prefer to treat them as constants and use local variables instead.

File: main.go [hunk 6]
```diff
@@ -309,7 +321,11 @@ func (p Ploop) Mount(target string, options map[string]string) (*flexvolume.Resp
 }
 
 func (p Ploop) Unmount(mount string) (*flexvolume.Response, error) {
-	if err := syscall.Unmount(mount, 0); err != nil {
+	err := syscall.Unmount(mount, 0)
+	if err == syscall.EINVAL {
+		//This isn't a mount point, continue and allow the ploop volume to be cleaned up on retry
```
Comment (dmission) [inline main.go]: It looks unrelated to secrets and readFromEnv - I suggest to make a separate PR for this with the use case described there.

**Review by dmission — APPROVED** [review]

**Comment by avagin** [comment]
@sanyu thank you!

**Comment by sanyu** [comment]
Thanks @avagin!
Would you mind having a look at https://github.com/virtuozzo/external-storage/pull/35 as well? This feature required changes in both, the controller and the flexdriver.
